driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -692,6 +692,10 @@ else if (fetchSize != Integer.MAX_VALUE)
    * and handle host failover.
    */
   void execute(final RequestHandler.Callback callback, final Statement statement) {
+    if (this.isClosed()) {
+      callback.onException(
+          null, new IllegalStateException("Could not send request, session is closed"), 0, 0);
+    }
```

Comment:
OK now the `isInit == true` branch is not covered in `executeAsync` because it runs the handler directly instead of delegating to this method.

How about we move the check to the beginning of the `RequestHandler` constructor? That way it will be done in a single place, and we don't have to worry about the different execution paths in the various session classes (which quite frankly are a mess at this point 😒 ).

Reply:
Yeah it looks like the change to call execute instead of new'ing up the `RequestHandler` is missing.   Agree, lets just move it to the constructor

Reply:
I ended implementing Andy's initial solution. Moving the check into the constructor caused other issues, namely how we would inform the caller of the constructor that there was an issue, and abort.

Everything should be routed through the execute method now, which should provide a single point to check. I also removed the superfluous check on the parent cluster.